lib/src/main/scala/code/chymyst/jc/Reaction.scala

```diff
@@ -264,5 +276,74 @@ final case class Reaction(info: ReactionInfo, body: ReactionBody, threadPool: Op
+      relevantMap.filterKeys( _ != molecule)
+    else
+      relevantMap.updated(molecule, valuesMap.updated(molValue, count - 1))
+  }
+
+  private[jc] def findInputMolecules(moleculesPresent: MoleculeBag): Option[(Reaction, InputMoleculeList)] = {
+    // Evaluate the static guard first.
+    val staticGuardPasses = info.guardPresence match {
+      case GuardPresent(_, Some(staticGuard), _) => staticGuard()
+      case _ => true
+    }
+    if (staticGuardPasses) {
+      val inputs = new InputMoleculeList(info.inputs.length)
+      // For each input molecule used by the reaction, find a random value of this molecule and evaluate the conditional.
+      val inputMoleculeInfos = info.inputsSorted
+      val inputMolecules = inputMoleculeInfos.map(_.molecule)
+
+      // Map of molecule values for molecules that are inputs to this reaction.
+      val relevantMap = moleculesPresent.getMap.filterKeys(m => inputMolecules.contains(m))
+
+      val found = inputMoleculeInfos.foldLeft[Iterable[(Map[Int, AbsMolValue[_]], BagMap)]](Iterable((Map(), relevantMap))) { (prev, inputInfo) =>
+        // `prev` contains the molecule value assignments we have found so far, as well as the molecule values that would remain in the soup after these previous molecule values were removed.
+        prev.flatMap {
+          case (prevValues, prevRelevantMap) =>
+            val valuesMap: Map[AbsMolValue[_], Int] = prevRelevantMap.getOrElse(inputInfo.molecule, Map())
+            val newFound = for {
+              newMolValue <- valuesMap.keysIterator // Do not eagerly evaluate the list of all possible values.
+              if inputInfo.admitsValue(newMolValue)
+              newRelevantMap = remove(prevRelevantMap, inputInfo.molecule, newMolValue)
+              newValues = prevValues.updated(inputInfo.index, newMolValue)
+            } yield (newValues, newRelevantMap)
+            newFound
+        }
+      }
+
+      // The reaction can run if `found` contains at least one admissible list of input molecules; just one is sufficient.
+
+      // Evaluate all cross-molecule guards: they must all pass on the chosen molecule values.
+      val filteredAfterCrossGuards = found.filter {
+        case (inputValues, _) => info.guardPresence match {
+          case GuardPresent(_, _, crossGuards) =>
+            crossGuards.forall {
+              case CrossMoleculeGuard(indices, _, cond) =>
+                cond.isDefinedAt(indices.flatMap(i => inputValues.get(i).map(_.getValue)).toList)
+            }
+          case _ => true
+        }
+      }
+
+      // Return result if found something. Assign the found molecule values into the `inputs` array.
+      filteredAfterCrossGuards.headOption.map {
+        case (inputValues, _) =>
+          inputValues.foreach {
+            case (i, molValue) =>
+              val mol = info.inputs(i).molecule
+              inputs.update(i, (mol, molValue))
+          }
+          (this, inputs)
+      }
+    } else None
```

Comment:
style: this else None is pretty far from its original if to recognize, not sure if you'd like to integrate with the initial match condition for staticGuardPasses.